### PR TITLE
Fix nested classes and unnamed package names for FunctionProcessor

### PR DESCRIPTION
**Pull request by andreaTP**
Turns out that `getQualifiedName` is not the proper identifier for the package name to be used

### CODE REVIEW

**Review by evacchi — APPROVED** [review]

File: function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java [hunk 1]
```diff
@@ -86,8 +96,13 @@ private void processModule(TypeElement type) {
...
             }
         }
-
-        var cu = new CompilationUnit(type.getEnclosingElement().toString());
+        var pkg = getPackageName(type);
+        var packageName = pkg.getQualifiedName().toString();
+        var cu = (pkg.isUnnamed()) ? new CompilationUnit() : new CompilationUnit(packageName);
+        if (!pkg.isUnnamed()) {
+            cu.setPackageDeclaration(packageName);
+            cu.addImport(type.getQualifiedName().toString());
```
Comment (electrum) [inline function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java:104]: Is this needed? Imports aren't needed for the same package.

Also, I don't see any changes in `BasicMathGenerated` or `SimpleGenerated`.

**Review by electrum — APPROVED** [review]
Nice! I should have tested these cases.

The nested class usage could be a good pattern where you need one or a few host functions.

File: function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java [hunk 2]
```diff
@@ -77,6 +79,14 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
         return false;
     }
 
+    private PackageElement getPackageName(Element element) {
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java:82]: ok

File: function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java (hunk 1 above)
Comment (andreaTP) [reply function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java:104]: It's part of the fix:
https://github.com/dylibso/chicory/pull/510/files#diff-bd5d5e9bbce4b5d75cb3f17f85ffc906bfdb7125dfa7e37722381bcbfd701f05R3

the import is added if it differs from the `packageName` (e.g. when using nested classes).

File: function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java [hunk 3]
```diff
@@ -86,8 +96,13 @@ private void processModule(TypeElement type) {
                 functions.add(processMethod(member, (ExecutableElement) member, moduleName));
             }
         }
-
-        var cu = new CompilationUnit(type.getEnclosingElement().toString());
+        var pkg = getPackageName(type);
+        var packageName = pkg.getQualifiedName().toString();
+        var cu = (pkg.isUnnamed()) ? new CompilationUnit() : new CompilationUnit(packageName);
+        if (!pkg.isUnnamed()) {
+            cu.setPackageDeclaration(packageName);
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply function-processor/src/main/java/com/dylibso/chicory/function/processor/FunctionProcessor.java:103]: it works well in combination with the following `addImport`: https://github.com/dylibso/chicory/pull/510#discussion_r1746202913

basically delegated to JavaParser the disambiguation.

**Comment by andreaTP** [comment]
Since this unblocks other work, I go ahead and merge it.
Any additional comment is welcome and I'll do the fixes in a follow up PR when necessary.

**Pull request merged**